python_interface/design/03_pyHorace_description.md

```diff
@@ -0,0 +1,269 @@
+(Thus this method will not work with an out-of-process Python interpreter).
+
+First when the `DataTypes` class encounters a Python callable (a function or class which can be called), instead of passing this to Matlab,
+it stores the function reference in a [module-global dictionary](https://github.com/mducle/hugo/blob/master/pyHorace/FunctionWrapper.py#L21),
+and then creates a Matlab [`pythonFunctionWrapper`](https://github.com/mducle/hugo/blob/master/src/pythonFunctionWrapper.m) object
+which stores the (string) key to the Python function.
+Whenever `call_method` encounters a `pythonFunctionWrapper` object,
+it [wraps](https://github.com/mducle/hugo/blob/master/src/check_wrapped_function.m) this in a call to the mex `call_python` function with the correct key.
+Thus it appears to the user that Matlab is calling a Python function transparently (but actually happens is that Matlab calls `call_python`).
+`call_python` then [retrieves](https://github.com/mducle/hugo/blob/master/src/call_python.cpp#L390) the Python function reference from the module-global `dict` 
+and [calls](https://github.com/mducle/hugo/blob/master/src/call_python.cpp#L413) that function with the arguments given to it.
+
+The majority of the code in `call_python` is for data conversion between Matlab and Python.
+This is because `mex` functions may only take and return *Matlab* data types (Python objects are not allowed).
+Similarly the CPython `PyObject_Call` function requires that inputs to the called functions are all of the `PyObject` type.
+Thus `call_python` converts Matlab types to their Python equivalent and vice versa as per this table:
+
+|      Matlab    |      Python         | Note |
+| -------------- | ------------------- | ---- |
+| numeric array  | `numpy.ndarray`     |      |
+| numeric array  | `list` of numbers   | p2m  |
+| `char` array   | `str`               |      |
+| `string`       | `str`               |      |
+| `string` array | `list` of `str`     |      |
+| `struct`       | `dict`              |      |
+| `struct` array | `list` of `dict`    |      |
+| `cell` array   | `list`              |      |
+| `cell` array   | `tuple`             | p2m  |
+
+"p2m" indicates that conversion only applies from Python to Matlab 
+(from a Python `list` of numbers to a Matlab numeric array, and from a python `tuple` to a Matlab `cell` array).
+Other conversions apply in both directions (e.g. Matlab cell arrays are always converted to a Python `list` and vice versa).
+The conversion from a Matlab array to a Python array involves no data copies - the raw memory pointed to by Matlab is wrapped in a numpy `array` object.
+This is possible because that memory is only used in the Python function to be executed and not more widely.
+The reverse conversion (from a `numpy` array to a Matlab array) involves a data copy
+because Matlab `mex` does not support a shared data wrapping over memory which is not allocated by Matlab.
```

Comment:
This is a bit unclear because it does not say who allocates the memory. If the receiving memory was allocated by matlab, would that work?

Reply:
Yes.

So there are two factory functions defined by the `mex` API: `CreateArray` and `CreateArrayFromBuffer`. `CreateArray` takes raw memory pointers or iterators and copies the data returning a Matlab `Array` object. `CreateArrayFromBuffer` in principle takes a smart pointer (`std::unique_ptr`) and steals the reference which means it doesn't do a copy. 

However, [Matlab only supports](https://uk.mathworks.com/matlabcentral/answers/514456-using-a-custom-deleter-in-a-buffer_ptr_t-in-arrayfactory-createarrayfrombuffer) the use case where you allocate the memory pointer by the `std::unique_ptr` using its own `CreateBuffer` function. So, if you allocate the buffer yourself and then use `CreateArrayFromBuffer` the code will compile and run but then will result in a segfault when Matlab tries to delete the memory.

This is because although everything is run within a single process and shares the same memory, the base Matlab interpreter, the mex file and the Python interpreters all have different heaps. In principle, `std::unique_ptr` allows you to overcome this by passing the pointer to the correct heap deallocator along with the pointer. However, for its own reasons Matlab does not honour this and overwrites the passed deallocator with its own function (which can only deallocate memory allocated in its own heap). To be fair it does `typedef` the `std::unique_ptr` to its own `buffer_ptr` type so you shouldn't expect it to obey the standard behaviour but it's still annoying...